v2/transport_auth.go

```diff
@@ -0,0 +1,160 @@
+package plugin
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"fmt"
+	"time"
+)
+
+const ServerTLSName = "gotify.home.arpa"
+
+func BuildPluginTLSName(moduleName string) string {
+	moduleNameHash := sha256.Sum256([]byte(moduleName))
+	hashHex := hex.EncodeToString(moduleNameHash[:])
+	return fmt.Sprintf("%s.plugins.gotify.home.arpa", hashHex)
+}
+
+type EphemeralTLSClient struct {
+	caCert    *x509.Certificate
+	caPriv    ed25519.PrivateKey
+	tlsConfig *tls.Config
+}
+
+func (s *EphemeralTLSClient) createCertPool() *x509.CertPool {
+	pool := x509.NewCertPool()
+	pool.AddCert(s.caCert)
+	return pool
+}
+
+func (s *EphemeralTLSClient) ServerTLSConfig() *tls.Config {
+	return &tls.Config{
+		ServerName: ServerTLSName,
+		ClientAuth: tls.RequireAndVerifyClientCert,
+		ClientCAs:  s.createCertPool(),
+	}
+}
+
+func (s *EphemeralTLSClient) ClientTLSConfig(moduleName string) *tls.Config {
+	return &tls.Config{
+		Certificates: []tls.Certificate{
+			{
+				Certificate: [][]byte{s.caCert.Raw},
+				PrivateKey:  s.caPriv,
+			},
+		},
+		RootCAs:    s.createCertPool(),
+		ServerName: BuildPluginTLSName(moduleName),
+	}
+}
+
+func (s *EphemeralTLSClient) CACert() *x509.Certificate {
+	return s.caCert
+}
+
+func (s *EphemeralTLSClient) SignCSR(dnsName string, csr *x509.CertificateRequest) ([]byte, error) {
+	if err := csr.CheckSignature(); err != nil {
+		return nil, err
+	}
+	certTemplate := &x509.Certificate{
+		BasicConstraintsValid: true,
+		Subject: pkix.Name{
+			CommonName: dnsName,
+		},
+		DNSNames: []string{
+			dnsName,
+		},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().Add(time.Hour * 24 * 365),
+		KeyUsage:  x509.KeyUsageDigitalSignature,
+		ExtKeyUsage: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageServerAuth,
+			x509.ExtKeyUsageClientAuth,
+		},
+		IsCA: false,
+	}
+	certBytes, err := x509.CreateCertificate(rand.Reader, certTemplate, s.caCert, csr.PublicKey, s.caPriv)
+	if err != nil {
+		return nil, err
+	}
+	return certBytes, nil
+}
+
+func (s *EphemeralTLSClient) SignPluginCSR(moduleName string, csr *x509.CertificateRequest) ([]byte, error) {
+	return s.SignCSR(BuildPluginTLSName(moduleName), csr)
+}
+
+func NewEphemeralTLSClient() (*EphemeralTLSClient, error) {
+	caPub, caPriv, err := ed25519.GenerateKey(rand.Reader)
```

Comment:
You previously linked the hashicorp plugin docs, they using mutual tls too and provide this plugin library https://github.com/hashicorp/go-plugin which has support for mtls and grpc. Have you considered using this? (just a question, not request to use this lib now)

Reply:
I took a quick peek and there are a couple things:

- They use the Go RPC not gRPC.. I think since our call interface is simple I don't see a reason to use an interface exclusive to Go.
- I think a key difficulty would be try to get HTTP reverse proxies working over the Server-to-Plugin connection and we might need abstraction breakouts here (raw gRPC has support for multiplexing over HTTP/2 listeners: https://pkg.go.dev/google.golang.org/grpc#Server.ServeHTTP which allows us to route `Webhooker` request in the same socket). Vault has their own abstractions over HTTP and they don't have this problem.

I have an authentication plugin you can checkout the code pattern for: https://github.com/eternal-flame-AD/vault-auth-emerg-yubiotp